listings can now be specified as an object

hrj opened the pull request:
for #127 

This doesn't even require API breakage. If the parameter is a string, it is interpreted as a URL (as before). Else it is treated as an object.

I have tested this and it works (with both string and object parameter).

jvilk commented:
Sorry for the late review.

Could you add a backend-specific test for XHR that creates an XHR using an object listing, and tests that it works? Specifically, you should test directory listings and `stat` on folders + files (+ nonexistent files).

Since `stat` triggers a server request for size, you'll need to make sure it's a legal listing. Just take a subset of the `listings.json` that we use for the full unit tests.

I believe all you have to do is create a new test file in the `test/tests/fs/XmlHttpRequest` folder (which you'll need to create). In that test, create a new XHRFS using a fake directory listing.

The test should take a backup of the existing rootFS before installing a new one, and re-instate it at the conclusion of the unit test. I've already done this in a unit test for the MFS here:

https://github.com/jvilk/BrowserFS/blob/master/test/tests/fs/MountableFileSystem/mounting.js#L94

Make sense? Let me know if you have any questions, or if this seems like too much work.

jvilk commented on src/backend/XmlHttpRequest.ts:
```diff
@@ -29,25 +29,32 @@ export default class XmlHttpRequest extends file_system.BaseFileSystem implement
-      throw new Error("Unable to find listing at URL: " + listingUrl);
+
+    let listing:Object = null;
+    if (typeof listingUrlOrObj == "string") {
```
Nit: Use `===`.

Super nit: I also like using parens with `typeof`, e.g. `typeof(listingUrlOrObj)`, as it reads less ambiguously (but that's optional).

jvilk commented on src/backend/XmlHttpRequest.ts:
```diff
@@ -29,25 +29,32 @@ export default class XmlHttpRequest extends file_system.BaseFileSystem implement
+    let listing:Object = null;
+    if (typeof listingUrlOrObj == "string") {
+      listing = this._requestFileSync(<string> listingUrlOrObj, 'json');
+      if (listing == null) {
```
Super nit: Use `!listing`. This will capture the `null` and `undefined` cases without resorting to a `==`.

This is also an optional change.

jvilk commented on src/backend/XmlHttpRequest.ts:
```diff
@@ -29,25 +29,32 @@ export default class XmlHttpRequest extends file_system.BaseFileSystem implement
+    if (typeof listingUrlOrObj == "string") {
+      listing = this._requestFileSync(<string> listingUrlOrObj, 'json');
+      if (listing == null) {
+        throw new Error("Unable to find listing at URL: " + listingUrlOrObj);
```
Super nit (optional): You can use a TypeScript template string here, btw, if you want.

```{ts}
throw new Error(`Unable to find listing at URL: ${istingUrlOrObj}`);
```

The result is the same either way.

jvilk commented on src/backend/XmlHttpRequest.ts:
```diff
@@ -29,25 +29,32 @@ export default class XmlHttpRequest extends file_system.BaseFileSystem implement
-    if (listing == null) {
-      throw new Error("Unable to find listing at URL: " + listingUrl);
+
+    let listing:Object = null;
```
Nit: You should have a space between the declaration and the type annotation, e.g.

`let listing: Object = null;`

jvilk commented:
I also added some small stylistic comments. The logic looks good. Now we just need that test to make sure I never break it, and we're golden for merger. :)

hrj commented:
I have fixed the nits; will now work on adding the test.

BTW, I am a nitpicker myself; please pick as many nits as you find :+1: :smile:

hrj commented:
> Since stat triggers a server request for size, you'll need to make sure it's a legal listing. Just take a subset of the listings.json that we use for the full unit tests.

Not sure which `listings.json` you had in mind; I can't find any in this repo.

Meanwhile, I will just cookup my own listings object.

jvilk commented:
The one XHR uses for unit testing:

https://github.com/jvilk/BrowserFS/blob/master/Gruntfile.js#L287

hrj commented:
I tried adding the test, but it is probably not cleaning up after it runs, so other tests are failing?

hrj commented:
I fixed it. The test for overlayfs was expecting the `fixtures` folder to have a particular layout, so I moved my test input file (49chars.txt) to a separate folder.

jvilk commented on src/backend/XmlHttpRequest.ts:
```diff
@@ -29,25 +29,32 @@ export default class XmlHttpRequest extends file_system.BaseFileSystem implement
     super();
-    if (listingUrl == null) {
-      listingUrl = 'index.json';
+    if (listingUrlOrObj == null) {
```
I used to do `==` as a shortcut for `=== null || === undefined`, but now use `!` as it's shorter & lints better.

No need to change if you don't feel like it.

jvilk replied:
(especially since this was my original code)

hrj replied:
No worries, I will fix it.

jvilk commented:
@hrj your test never runs. The folder has to be titled `XmlHttpRequest` (the name of the fs type).

It also has many problems, which you'll see when you try to run it. I'll comment on some.

jvilk commented on test/tests/fs/xhrFS/listing.js:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Unit tests for MountableFileSystem's mount/unmount features.
```
Should be updated.

jvilk commented on test/tests/fs/xhrFS/listing.js:
```diff
@@ -0,0 +1,65 @@
+module.exports = function() {
+  var oldRootFS = fs.getRootFS();
+
+  var rootForXFS = new BrowserFS.FileSystem.InMemory();
```
Why create an in-memory FS? You should initialize the root to be the XHRFS you create w/ the listings object.

hrj replied:
I was following the mountable-fs template, without really understanding what that test was about. Will fix it.

jvilk commented on test/tests/fs/xhrFS/listing.js:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  var newXFS = new BrowserFS.FileSystem.XmlHttpRequest(listing, "test/tests/fs/xhrFS");
+  rootForXFS.mount("/tRoot", newXFS);
```
InMemoryFS doesn't have a `mount`. You should just make `newXFS` the root with `initialize` instead.

jvilk commented:
Hope that helps! We're close to landing this one. Just gotta make the test run & pass.

hrj commented:
All green now!

Sorry for the commit spam. This is harder than it needs to be because the tests fail locally for me, even on the master branch. (local firefox version is 43 v/s 31 in travis)

jvilk commented:
Wait, really? They fail? I have Firefox 43 installed locally, and it works here (on Mac). What OS?

hrj commented:
OS is ubuntu. The failure looks like this:

```
Firefox 43.0.0 (Ubuntu 0.0.0) BrowserFS Tests General Tests std-streams.js [BufferCoreArray] FAILED
	0 == 6
	wrapObject/</wrappedObj[name]@/home/ubuntu/dopp/BrowserFS/test/harness/test.js:5462:28
	[133]</</module.exports/<@/home/ubuntu/dopp/BrowserFS/test/harness/test.js:21526:5
	[16]</EventEmitter.prototype.emit@/home/ubuntu/dopp/BrowserFS/test/harness/test.js:5154:9
	[13]</</Process</Process.prototype.exit@/home/ubuntu/dopp/BrowserFS/test/harness/test.js:4796:9
	defineKey/processProxy[key]@/home/ubuntu/dopp/BrowserFS/test/harness/test.js:4579:20
	generateTest/</<@/home/ubuntu/dopp/BrowserFS/test/harness/test.js:15994:21
	waitsFor/interval<@/home/ubuntu/dopp/BrowserFS/test/harness/test.js:15969:13

```
